lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/BuildCriOauthRequestHandler.java

```diff
@@ -333,8 +345,41 @@ private JWEObject signEncryptJar(
                         evidenceRequest,
                         context);
 
-        RSAEncrypter rsaEncrypter = new RSAEncrypter(oauthCriConfig.getParsedEncryptionKey());
-        return AuthorizationRequestHelper.createJweObject(rsaEncrypter, signedJWT);
+        RSAKey encKey;
+        if (cachedOAuthCriEncryptionKeys != null
+                && !cachedOAuthCriEncryptionKeys.isEmpty()
+                && cachedOAuthCriEncryptionKeys.get(cri.getId()) != null
+                && !cachedOAuthCriEncryptionKeys.get(cri.getId()).isExpired()) {
+            encKey = cachedOAuthCriEncryptionKeys.get(cri.getId()).key();
+        } else {
+            encKey = oAuthKeyService.getValidEncryptionKey(oauthCriConfig);
+
+            var cacheDuration =
+                    configService.getParameter(ConfigurationVariable.OAUTH_KEY_CACHE_DURATION_MINS)
+                                    == null
+                            ? DEFAULT_OAUTH_CRI_ENC_KEY_CACHE_DURATION_MINUTES
+                            : Integer.parseInt(
+                                    configService.getParameter(
+                                            ConfigurationVariable.OAUTH_KEY_CACHE_DURATION_MINS));
+            cacheEncryptionKey(encKey, cri, cacheDuration);
+        }
+
+        RSAEncrypter rsaEncrypter = new RSAEncrypter(encKey);
+        return AuthorizationRequestHelper.createJweObject(
+                rsaEncrypter, signedJWT, encKey.getKeyID());
+    }
+
+    private static void cacheEncryptionKey(RSAKey key, Cri cri, Integer cacheDuration) {
+        // Initialise cachedOAuthCriEncryptionKeys if still null
+        if (cachedOAuthCriEncryptionKeys == null) {
+            cachedOAuthCriEncryptionKeys = new HashMap<>();
+        }
+        // Cache the key in a variable outside of the handler so it exists in
+        // the Lambda's memory and is still accessible between invocations that
+        // occur at short intervals from one another
+        var expiryDate = LocalDateTime.now().plusMinutes(cacheDuration);
+        cachedOAuthCriEncryptionKeys.put(
+                cri.getId(), new CachedOAuthCriEncryptionKey(key, expiryDate));
```

Comment:
I think you need to key this on CRI _and_ connection ID (the `connection` variable in the handle method) - e.g. in staging we might be using either the stub or real CRI, depending on a feature flag (and these will have different endpoints).

Reply:
Alternatively, if you move the logic into `OAuthKeyService`, you could cache on the JWKS URL instead